lib/optimizer/src/Transformer/PreloadHeroImage.php

```diff
@@ -0,0 +1,488 @@
+        }
+
+        return null;
+    }
+
+    /**
+     * Detect a hero image candidate from an <amp-img> element.
+     *
+     * @param DOMElement $element Element to detect for.
+     * @return HeroImage|null Detected hero image candidate, or null if none detected.
+     */
+    private function detectHeroImageCandidateForAmpImg(DOMElement $element)
+    {
+        $src = $element->getAttribute(Attribute::SRC);
+
+        if (empty($src)) {
+            return null;
+        }
+
+        if (! Url::isValidImageSrc($src)) {
+            return null;
+        }
+
+        if ((new ImageDimensions($element))->isTiny()) {
+            return null;
+        }
+
+        $srcset = $element->getAttribute(Attribute::SRCSET);
+        $media  = $element->getAttribute(Attribute::MEDIA);
+
+        return new HeroImage($src, $media, $srcset, $element);
+    }
+
+    /**
+     * Detect a hero image candidate from a video's poster (= placeholder) image.
+     *
+     * @param DOMElement $element Element to detect for.
+     * @return HeroImage|null Detected hero image candidate, or null if none detected.
+     */
+    private function detectHeroImageCandidateForPosterImage(DOMElement $element)
+    {
+        $poster = $element->getAttribute(Attribute::POSTER);
+
+        if (! $poster) {
+            return null;
+        }
+
+        if (! Url::isValidImageSrc($poster)) {
+            return null;
+        }
+
+        if ((new ImageDimensions($element))->isTiny()) {
+            return null;
+        }
+
+        $media = $element->getAttribute(Attribute::MEDIA);
+
+        return new HeroImage($poster, $media, '');
+    }
+
+    /**
+     * Detect a hero image candidate from an iframe's placeholder image.
+     *
+     * @param DOMElement $element Element to detect for.
+     * @return HeroImage|null Detected hero image candidate, or null if none detected.
+     */
+    private function detectHeroImageCandidateForIframePlaceholderImage(DOMElement $element)
+    {
+        // A placeholder <amp-img> is required to preload an image for an iframe.
+        if (! $element->hasChildNodes()) {
+            return null;
+        }
+
+        if ((new ImageDimensions($element))->isTiny()) {
+            return null;
+        }
+
+        return $this->getPlaceholderImage($element);
+    }
+
+    /**
+     * Get the placeholder image for a given element.
+     *
+     * @param DOMElement $element Element to check the placeholder image for.
+     * @return HeroImage|null Placeholder image to use or null if none found.
+     */
+    private function getPlaceholderImage(DOMElement $element)
+    {
+        foreach ($element->childNodes as $childNode) {
+            if (
+                ! $childNode instanceof DOMElement
+                || $childNode->tagName !== Extension::IMAGE
+                || ! $childNode->hasAttribute(Attribute::PLACEHOLDER)
+            ) {
+                continue;
+            }
+
```

Comment:
Some new logic can be added here to see if `$childNode->tagName === Extension::IMAGE`, and if so, use its `$src`. Otherwise, query `$childNode` for the first descendant `Extension::IMAGE` and use it instead.

Reply:
Nevertheless, should it be checking for `amp-anim` in addition to `amp-img`?

Reply:
I think it makes sense to check whether the first descendant `Extension::IMAGE` actually has the `placeholder` attribute.

Otherwise, we might preload a `fallback` image instead, which is not supposed to be loaded at all unless something is broken.

Reply:
Oh, we're already within a placeholder here, so I guess that check is futile. In that case, I'm not sure why we're looking for descendants. What would the placeholder be nested in?

Reply:
Ok, I assume with the first descendant `Extension::IMAGE`, you're referring to the case where you have arbitrary HTML as the fallback, instead of a singular `Extension::IMAGE`.

I noticed that we can preload a regular `<img data-hero>` as well, provided we skip the SSR generation of an additional `<img>` tag. This would be true for the above as well.

Reply:
@westonruter Is this what you had in mind: https://github.com/ampproject/amp-wp/pull/5350/commits/dda3065da2b9412fa84afbd4ba2d07fc57a98549 ?

Reply:
Yes, mostly. Although instead of a `<p>` the more common case would be for the `placeholder` to be a `<a>` link. See:

https://github.com/ampproject/amp-wp/blob/31758e96a656e2cf722cce842bb301d7a6152763/tests/php/test-class-amp-youtube-embed-handler.php#L152

Reply:
This is the list of components I'm currently considering as "hero image material":
https://github.com/ampproject/amp-wp/pull/5350/commits/6ae3e3ee5e20676012e7b5674c0f5a08f66ee736#diff-6d41b4074203a337a3ed777fee7549bd417fc11c40b8379d75a38d616126709cR83-R101

Reply:
Why limit it to those AMP components? Does it hurt to just look for _any_ placeholder with an image of any AMP component?

Reply:
This is also used to guess what the hero image could be, in the hopes we do indeed guess correctly and optimize what will end up being the LCP.

If we accept any AMP components, the risk is higher of optimizing something that will not be the LCP. Therefore, I only kept elements which are useful as an intro to an article or page and that usually produce a rather big render target.

Some AMP components are most of the time small or not even being displayed, or they only make sense in the footer of the page (below the fold). Technically, we could still preload these, but that will hurt the performance wrt LCP instead.